metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1016,6 +1042,55 @@ ApiError electLeader(String topic, int partitionId, ElectionType electionType,
         return ControllerResult.of(records, null);
     }
 
+    boolean arePartitionLeadersImbalanced() {
+        return !imbalancedPartitions.isEmpty();
+    }
+
+    /**
+     * Attempt to elect a preferred leader for all topic partitions that a leader that is not the preferred replica.
+     *
+     * The response() method in the return object is true if this method returned without electing all possible preferred replicas.
+     * The quorum controlller should reschedule this operation immediately if it is true.
+     *
+     * @return All of the election records and if there may be more available preferred replicas to elect as leader
+     */
+    ControllerResult<Boolean> maybeBalancePartitionLeaders() {
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        boolean rescheduleImmidiately = false;
+        for (TopicIdPartition topicPartition : imbalancedPartitions) {
+            if (records.size() >= maxElectionsPerImbalance) {
+                rescheduleImmidiately = true;
+                break;
+            }
+
+            TopicControlInfo topic = topics.get(topicPartition.topicId());
+            if (topic == null) {
+                log.error("Skipping unknown imbalanced topic {}", topicPartition);
+                continue;
+            }
+
+            PartitionRegistration partition = topic.parts.get(topicPartition.partitionId());
+            if (partition == null) {
+                log.error("Skipping unknown imbalanced partition {}", topicPartition);
+                continue;
+            }
```

Comment:
Hmm, I think if we hit one of these cases (which seems unlikely), we'll never remove the item from `imbalancedPartitions` since we won't get a PartitionChangeRecord. I'm not really sure how we can properly handle this though since we don't want to mutate `imbalancedPartitions` outside of a "relay" method. WDYT?

Reply:
This case means that there is a bug in the controller implementation. We can either skip the topic partition or renounce the active controller. I decided to skip it for now.